test/unit3/test_random.py

```diff
@@ -108,6 +109,365 @@ def test_sample_all(self):
+        with self.assertRaises(TypeError):
+            choices(data, cum_weights=10, k=5)                           # non-iterable cum_weights
+        with self.assertRaises(TypeError):
+            choices(data, cum_weights=[None]*4, k=5)                     # non-numeric cum_weights
+        with self.assertRaises(TypeError):
+            choices(data, range(4), cum_weights=range(4), k=5)           # both weights and cum_weights
+        for weights in [
+                [15, 10, 25, 30],                                                 # integer cum_weights
+                [15.1, 10.2, 25.2, 30.3],                                         # float cum_weights
+                # fractions module isn't implemented [Fraction(1, 3), Fraction(2, 6), Fraction(3, 6), Fraction(4, 6)], # fractional cum_weights
+        ]:
+            self.assertTrue(set(choices(data, cum_weights=weights, k=5)) <= set(data))
+
+        # Test weight focused on a single element of the population
+        self.assertEqual(choices('abcd', [1, 0, 0, 0]), ['a'])
+        self.assertEqual(choices('abcd', [0, 1, 0, 0]), ['b'])
+        self.assertEqual(choices('abcd', [0, 0, 1, 0]), ['c'])
+        self.assertEqual(choices('abcd', [0, 0, 0, 1]), ['d'])
+
+        # Test consistency with random.choice() for empty population
+        with self.assertRaises(IndexError):
+            choices([], k=1)
+        with self.assertRaises(IndexError):
+            choices([], weights=[], k=1)
+        with self.assertRaises(IndexError):
+            choices([], cum_weights=[], k=5)
+
+    def test_choices_subnormal(self):
+        # Subnormal weights would occasionally trigger an IndexError
+        # in choices() when the value returned by random() was large
+        # enough to make `random() * total` round up to the total.
+        # See https://bugs.python.org/msg275594 for more detail.
+        choices = random.choices
+        choices(population=[1, 2], weights=[1e-323, 1e-323], k=5000)
+
+    def test_choices_with_all_zero_weights(self):
+        # See issue #38881
+        with self.assertRaises(ValueError):
+            random.choices('AB', [0.0, 0.0])
+
+    def test_choices_negative_total(self):
+        with self.assertRaises(ValueError):
+            random.choices('ABC', [3, -5, 1])
+
+    def test_choices_infinite_total(self):
+        with self.assertRaises(ValueError):
+            random.choices('A', [float('inf')])
+        with self.assertRaises(ValueError):
+            random.choices('AB', [0.0, float('inf')])
+        with self.assertRaises(ValueError):
+            random.choices('AB', [-float('inf'), 123])
+        with self.assertRaises(ValueError):
+            random.choices('AB', [0.0, float('nan')])
+        with self.assertRaises(ValueError):
+            random.choices('AB', [float('-inf'), float('inf')])
+class TestBasicOps:
+    # Superclass with tests common to all generators.
```

Comment:
Thanks for adding these
can we move TestBasicOps to the top of the file to better match the CPython test file
(makes future edits easier)
(and similar for MersenneTwister_TestBasicOps)